File: python/bitree.py
"""
NAME 
    bitree - binary tree algorithms

DESCRIPTION
    This module implements a few tree-related algorithms. 

    * bfs         breadth-first search
    -----------------------------------------
    * preorder    iterative dfs in  pre-order 
    * inorder     iterative dfs in   in-order 
    * postorder   iterative dfs in post-order 

FUNCTIONS
    bfs(root)
        Return nodes of a binary tree in level-order.

    preorder(root)
        Return nodes of a binary tree in pre-order.

    inorder(root)
        Return nodes of a binary tree in in-order.

    postorder(root)
        Return nodes of a binary tree in post-order.
"""

from collections import deque

class TreeNode:
    def __init__(self, value, left=None, right=None):
        self.val = value
        self.left = left
        self.right = right


def bfs(root):
    """Return nodes of a binary tree in level-order."""
    ans = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            ans.append(node)
            queue.append(node.left)
            queue.append(node.right)
    return ans 


def dfs(root):
    """Recursively depth-first traverse a binary tree."""
    #   inorder traversal : left-node-right
    #  preorder traversal : node-left-right
    # postorder traversal : left-right-node
    if root: 
        yield from dfs(root.left)
        yield root.val
        yield from dfs(root.right)


def preorder(root):
    """Return nodes of a binary tree in pre-order."""
    ans = []
    stack = [root]
    while stack:
        node = stack.pop()
        if node: 
            ans.append(node)
            stack.append(node.right)
            stack.append(node.left)
    return ans 


def inorder(root):
    """Return nodes of a binary tree in in-order."""
    ans = []
    node = root 
    stack = []
    while node or stack:
        if node: 
            stack.append(node)
            node = node.left 
        else: 
            node = stack.pop() 
            ans.append(node)
            node = node.right 
    return ans 


def postorder(root):
    """Return nodes of a binary tree in post-order."""
    ans = []
    node, prev = root, None
    stack = []
    while node or stack:
        if node: 
            stack.append(node)
            node = node.left 
        else: 
            node = stack[-1] 
            if node.right and node.right != prev: node = node.right 
            else: 
                ans.append(node) 
                stack.pop() 
                prev = node 
                node = None
    return ans 

File: ruby/bitree.rb
module CPLib

  ##
  # This class provides key functionalities related to binary tree such as
  # preorder, inorder and postorder traversals.
  class BiTree

    ##
    # Traverse a binary tree rooted at +node+ in preorder.
    def self.preorder(node)
      ans = []
      stack = [node]
      while stack.any?
        node = stack.pop
        ans << node.val
        stack << node.right if node.right
        stack << node.left if node.left
      end
      ans
    end

    ##
    # Traverse a binary tree rooted at +node+ in inorder.
    def self.inorder(node)
      ans = []
      stack = []
      while node || stack.any?
        if node
          stack << node
          node = node.left
        else
          node = stack.pop
          ans << node.val
          node = node.right
        end
      end
      ans
    end

    ##
    # Traverse a binary tree rooted at +node+ in postorder.
    def self.postorder(node)
      ans = []
      prev, stack = nil, []
      while node || stack.any?
        if node
          stack << node
          node = node.left
        else
          node = stack[-1]
          if node.right && node.right != prev
            node = node.right
          else
            ans << node.val
            stack.pop
            prev, node = node, nil
          end
        end
      end
      ans
  end

end
